### DISCUSSION TITLE
Follow up on changes from #2419 (layers shift)

### DISCUSSION THREAD

**Discussion by tgriesser** [urql-graphql/urql#2440] [Q&A]
👋 Following up with a separate discussion thread on the changes in #2419 

For background, we're using `urql` / `graphcache` / `@urql/vue` in the new [Cypress v10](https://github.com/cypress-io/cypress/pull/18924) which is due to be released within the next month, and it's been great to work with, so thank you.

Still trying to pin down the root cause of how these changes are interacting with some custom things we're doing. Can follow up another time with a better show-and-tell of how we're using urql/graphcache on both the client and in the electron process as a server-side cache for our remote stitched GraphQL API (which I won't go into here, since it's unrelated to this issue).

---

There are two places where we're seeing the recent changes causing differences in behavior. The first one I'm slightly more concerned about than the second, because it seems like a query is executed a few times and is potentially resolving with the wrong value from the cache, unrelated to a subscription. The second case is funny because it basically designed to force the unexpected race situation you mentioned between the socket & query resolution 😄

Can probably work around both issues, but would like to better understand *why* before attempting to do so. Also, I apologize in advance for the lack of a simple repo here, I have gotten some good debug logs, which I think helps at least isolate some of the behavior the issue.

### The first issue:

The following logs are coming from a custom debug exchange <a href="#custom-debug-exchange">[1]</a>, which normalizes the duplicated data a bit to make the logs decipherable.

Here is the diff (4.3.6 on the left, 4.4.0 on the right): https://www.diffchecker.com/OmarUwtw
The raw data used for the diff: https://gist.github.com/tgriesser/172e31605fdd8d64acd26f2d442d2405

So what we see here is a whole bunch of queries that are triggered repeatedly. To explain that, we have a custom exchange <a href="#pub-sub-exchange">[2]</a> which is modeled after the retry-exchange, which tracks the currently mounted queries to allow the backend to signal down via websocket to "refetch". It's a pretty heavy-handed way of keeping state up to date, which came about through incremental migration from the standard document `cacheExchange` and a lack of proper subscriptions to manage the updates at the time. We use it more than we should, but querying against a local (electron process) GraphQL api is cheap and doing it right with our newer subscriptions layer has been lower on the todo list than other things.

When the Cypress app is starting up, after you choose a "testing type" (`End-to-End` or `Component`), it runs some code, sources some config, spawns a child process, does some communication, and then hits a settled state. Each time the state transitions, we're triggering the signal to refresh queries on the client, which is why we see so many of the same queries re-executed multiple times here.

You'll notice a "data" value in the diff log, which is a hashed stable-stringify of the data payload that comes back (you can see the full values at the bottom of the diff), but a quick summary of what's different about each.

```
6321d3497fa752a3142ba36b90408613 = loadingConfigFile: false, loadingNodeEvents: true  -- loading screen
85e324c419dc3b3b358a7e950841d06d = loadingConfigFile: true, loadingNodeEvents: true  -- loading screen
97f38e4d9ec72db719593e767c5381b2 = loadingConfigFile: false, loadingNodeEvents: false   -- settled
```

So the thing that stands out as being odd, on line `467`, the "settled" value which we use to indicate we're ready to progress to the next page is resolved on both sides:

<img width="1439" alt="image" src="https://user-images.githubusercontent.com/154748/167504562-efedb3a1-c3e8-4500-b357-0595df584afd.png">

Then, when things to diverge, the same query is executed & resolved as a cache-hit, but in 4.4.0 it is using a value it resolved with from a much earlier execution of the query, which contains a state that shows a loading screen:

<img width="1491" alt="image" src="https://user-images.githubusercontent.com/154748/167520278-81af0869-fc15-48eb-beea-f5ecd4997ce7.png">

The data it's resolving is from earlier than the "settled" `97f38e4d9ec72db719593e767c5381b2` resolved value for the `MainLaunchpadQuery` query, so it almost looks like it's somehow resolving with the wrong data.

So the question is - did something change here with the layer shifting that would cause it to be resolving with stale data?

### The second issue

@kitten quote from #2419

> However, typically I wouldn't expect this to cause an issue since it's very unlikely for a query to be completing or updating with new data while a subscription is updating with a new event. In other words, what's the missing piece here is why the order would be so sensitive and unstable when the query isn't realtime

I had initially played around with `@defer` and wasn't really happy with the tradeoffs relative to what we're after for our use-case, so I decided to try a different concept, sort of a server-controlled "defer".

Consider the following query:

```graphql
query {
  cloudViewer {  # remote stitched
    id
    organizations(first: 10) {
      nodes {
        id
        name
      }
    }
  }
  currentProject {
    id
    path
    config
    cloudProject {  # remote stitched
      id
      name
      runs(first: 10) {
        nodes {
          id
        }
      }
    }
  }
}
```

There are two pieces of data that normally resolve from a remote API. Because it's a local electron app, we don't want these to delay resolving the query, so if we don't already have a resolved value in the server urql cache, we resolve immediately with "null":

```graphql
{
  cloudViewer: null,
  currentProject: {
    id: "...",
    path: "...",
    config: "...",
    cloudProject: null
  }
}
```

Essentially, we want to be able to stitch in remote fields and have it feel like a natural extension of the API client-side, and have the server manage the resolution of the data into the cache as it comes in, or resolve the field immediately if we have all of the data needed in the server-side cache.

As we resolve the remote data in, we send a custom `FragmentDefinitionNode` based on the `info.fieldNodes` to be able to "push" the values down to the client via a subscription, which in-turn calls `cache.writeFragment` with the fragment & data:
```ts
updates: {
  Subscription: {
    pushFragment (parent, args, cache, info) {
      const { pushFragment } = parent as { pushFragment: { id?: string, fragment: DocumentNode, data: any, typename: string }[] }

      for (const toPush of pushFragment) {
        cache.writeFragment(toPush.fragment, toPush.data)
      }
    },
  },
}
```

```ts
{
  pushFragment: {
    fragment: `
      fragment GeneratedFragment on Query {
        cloudViewer {
          id
          organizations(first: 10) {
            nodes {
              id
              name
            }
          }          
        }
      }
    `,
    data: {
      "cloudViewer": "..."
    }
  }
}

{
  pushFragment: {
    fragment: `
      fragment GeneratedFragment on CurrentProject {
        id
        cloudProject {
          id
          name
          runs(first: 10) {
            nodes {
              id
            }
          }
        }
      }
    `,
    data: {
      "id": "...",
      "cloudProject": "..."
    }
  }
}
```

So what this comes down to, as far as I can guess, is that in our local tests, the fragment is being pushed via the subscription before the query it's meant to resolve "into" is resolved. This is likely only happening because of our tests / local env forcing this behavior with a quicker resolution of data than would normally be expected (I think?).

Could probably get around this with something like waiting a delay after the original query's `req` headers are sent before sending the subscription, but I was hoping to be able to rely on the previous guarantees urql seemed to provide. Also was interested to see if you had any feedback on this idea and if there was anything I might be overlooking using this approach.

One other super minor thing while I'm here. Any chance you've ever considered shipping with a conditional "development" build of the Urql suite, similar to what React does? Even with the sourcemaps, terser basically does too good of a job with the minification to do any sort of reasonable breakpoint debugging. I find myself `yarn link`'ing to debug, but it's a bit of a cumbersome workflow.

Thanks for any help with this!

----

#### Custom debug exchange

<div id="custom-debug-exchange" />

```ts
window.__toPrint__ = {
  logs: [],
  queries: {},
  data: {},
}

const detailedDebugExchange: Exchange = ({ forward }) => {
  return (ops$) => {
    return pipe(
      ops$,
      tap((op) => {
        window.__toPrint__.queries[(op.query.definitions[0] as OperationDefinitionNode).name?.value] = print(op.query)
        window.__toPrint__.logs.push(JSON.parse(JSON.stringify({
          debugType: 'incoming',
          timestamp: new Date().valueOf(),
          ...op,
          query: (op.query.definitions[0] as OperationDefinitionNode).name?.value,
        })))
      }),
      forward,
      tap((result) => {
        const hash = md5(jsonStableStringify(result.data))
        window.__toPrint__.data[hash] = result.data
        window.__toPrint__.logs.push(JSON.parse(JSON.stringify({
          debugType: 'completed',
          timestamp: new Date().valueOf(),
          ...result,
          data: hash,
          operation: {
            ...result.operation,
            query: (result.operation.query.definitions[0] as OperationDefinitionNode).name?.value,
          },
        })))
      }),
    )
  }
}
```

#### Pub-sub exchange

<div id="pub-sub-exchange" />

```ts
export const pubSubExchange = (io: Socket): Exchange => {
  return ({ client, forward }) => {
    const watchedOperations = new Map<number, Operation>()
    const observedOperations = new Set<number>()
    function reexecuteOperation (op: Operation) {
      client.reexecuteOperation(
        client.createRequestOperation('query', op, {
          requestPolicy: 'network-only',
          fetchOptions: { headers: { 'x-cypress-graphql-refetch': 'true' } },
        }),
      )
    }
    io.on('graphql-refetch', () => {
        watchedOperations.forEach((op) => reexecuteOperation(op))
    })
    const processIncomingOperation = (op: Operation) => {
      if (op.kind === 'teardown' && observedOperations.has(op.key)) {
        watchedOperations.delete(op.key)
        observedOperations.delete(op.key)
      }
    }
    const processResultOperation = (op: OperationResult) => {
      if (op.operation.kind === 'query' && !observedOperations.has(op.operation.key)) {
        watchedOperations.set(op.operation.key, op.operation)
        observedOperations.add(op.operation.key)
      }
    }
    return (ops$) => {
      return pipe(forward(pipe(ops$, tap(processIncomingOperation))), tap(processResultOperation))
    }
  }
}
```

**Comment by kitten** [comment]
sorry for not getting to this sooner! I'm hoping actually that we can somehow codify this into tests and then reason about whether the outcome make sense or not and determine whether it's consistent and expected behaviour or not. 🤔

> but in 4.4.0 it is using a value it resolved with from a much earlier execution of the query

This is certainly what throws me off most. The query has finished and hence we should see the latest value. This could be a deduplication issue or an ordering issue. But the important note here is that when a `network-only` result comes back, the result will always reflect its layer and no layers on top. This means that we may have an older layer on top that's for some reason not removed.
That would mean though that we'd have at least three layers of overlapping data, I suppose, where one is still pending.

The second case sounds exactly like what you're assuming. The subscription will _not_ be pending and when the query is still pending its results end up below the queries. The fix here could be a multitude of things. In the past, we've also recommended that a solution may be to issue "fake queries".
However, I can see how that's less practical than live queries and such. I don't have a great answer for this problem yet. But I believe this is expected behaviour because we can't tell which result is more recent automatically.

Edit on the above: I believe we actually had an escape hatch for data set to `undefined` before. In essence, it'd behave just like `@defer` but without waiting for future results via `hasNext` on responses... Not sure whether that's still intact though. It may have also been a schema awareness thing only. Are you using schema awareness?

I believe the best thing forward would be to _only_ disable the layer shifting via a flag and then see whether this resolves all problems. If not we can first narrow this down further.

We do actually have preview builds via GitHub Actions on each PR (including the "Version" PR for releases), however, the artifacts are zipped tarballs, so without us writing a lambda (and GitHub's API is painful to deal with imho) they can't be used directly without downloading and unzipping the tarball manually first.

**Comment by tgriesser** [comment]
> sorry for not getting to this sooner!

No need to apologize, I feel like your response here was very fast, makes me feel like a bad maintainer on my own projects 😅 

> This is certainly what throws me off most. The query has finished and hence we should see the latest value. This could be a deduplication issue or an ordering issue. But the important note here is that when a network-only result comes back, the result will always reflect its layer and no layers on top. This means that we may have an older layer on top that's for some reason not removed.
That would mean though that we'd have at least three layers of overlapping data, I suppose, where one is still pending.

Yeah, it sounds like this might be what's happening. In breaking down the operation log captured here one-by-one & looking at only the resolutions of the `MainLaunchpadQuery` (key `2691033737`), we can see this actually happening twice:

1. Resolves `cache-first` miss with `6f0cb10f6451ef122cf06180da1ad4a9` in both (expected) ✅ 

<img width="1361" alt="image" src="https://user-images.githubusercontent.com/154748/167856584-fa3de5d8-c9b6-40b9-a924-8984c7a3fa2e.png">

2. Resolves `network-only` with `85e324c419dc3b3b358a7e950841d06d` in both (expected) ✅

<img width="1397" alt="image" src="https://user-images.githubusercontent.com/154748/167856695-82b100bc-8e6f-4703-95c3-ea745f486f73.png">

3. Resolves `cache-first` (of `cache-and-network`) with `85e324c419dc3b3b358a7e950841d06d` in both (expected) ✅

<img width="1384" alt="image" src="https://user-images.githubusercontent.com/154748/167857187-418f9999-f5b2-49c9-bec9-9f21c0a4416a.png">

4. Resolves `network-only` (from `cache-and-network`) with `6321d3497fa752a3142ba36b90408613` in both (expected) ✅

<img width="1350" alt="image" src="https://user-images.githubusercontent.com/154748/167857442-6b84f143-ea3a-4877-b72d-0a328575f84c.png">

5. Resolves `cache-first` (from `cache-and-network`) with `6321d3497fa752a3142ba36b90408613` in 4.3.6 (expected), and `85e324c419dc3b3b358a7e950841d06d` in 4.4.0 (unexpected) ❌, that was the resolved value from step 2+3, and no other payloads have been different in other requests such that the cache would be inconsistent

<img width="1346" alt="image" src="https://user-images.githubusercontent.com/154748/167857634-4af1956f-4683-4717-bfe2-9f2bcaba9a6d.png">

6. Resolves `network-only` (from `cache-and-network`) with `97f38e4d9ec72db719593e767c5381b2` in both (back to expected state) ✅

<img width="1388" alt="image" src="https://user-images.githubusercontent.com/154748/167858783-f2cc8408-b1da-4432-a63f-650ed3780c66.png">

7. Resolves `cache-first` from (cache-and-network) with `97f38e4d9ec72db719593e767c5381b2` in 4.3.6 (expected), and `85e324c419dc3b3b358a7e950841d06d` in 4.4.0 (unexpected) ❌, again, the same unexpected resolved value from step 2+3, and again, no other payloads have been different in other requests such that the cache would be inconsistent

<img width="1386" alt="image" src="https://user-images.githubusercontent.com/154748/167858967-c13b7169-14a1-4f39-a89b-1865df660ed9.png">

So in both cases, it seems `cache-first` is resolving incorrectly with a the same "stale" layer.

I guess the only other thing that's worth noting is that there are a few subscriptions that are mounted around that time, and one that has a value resolved in while the initial query is in-flight:

<img width="1412" alt="image" src="https://user-images.githubusercontent.com/154748/167861734-c69eeff9-20d4-48aa-b94b-5165cf23604e.png">

Also, was looking closer at the change in #2419, I was at first wondering about the early `return`, and the fact that previously `commutativeKeys.add` was always called, so I added:

```ts
} else if (!hasNext) {
  data.commutativeKeys.add(layerKey);
  return;
}
```

and no urql unit tests broke, which make me think the tests for this behavior aren't exhaustive. That change didn't seem to fix the issue I'm seeing though. 

Was also doing a bit of debugging around the `for` loop & `optimisticOrder` splicing which seems to have changed a bit - nothing conclusive yet, still tracing through the different codepaths to understand where something unexpected might be happening.

> Edit on the above: I believe we actually had an escape hatch for data set to undefined before. In essence, it'd behave just like @defer but without waiting for future results via hasNext on responses... Not sure whether that's still intact though. It may have also been a schema awareness thing only. Are you using schema awareness?

We are using schema awareness, when you say "fake queries", do you mean like the subscription would come down with a value, and then we'd use that to some sort of a query with it so it layers properly? And when you say "data set to undefined", is that the `data` response from the request, or something else?

> I believe the best thing forward would be to only disable the layer shifting via a flag and then see whether this resolves all problems. If not we can first narrow this down further.

Is this something I should do on my side, or how were you anticipating this would work?

**Reply by JoviDeCroock** [reply]
Hey @tgriesser what do you think about taking the time to pair on this? 😅 it's hard for me to go through all the data in that gist but I do think that there's the possibility of some events arriving in an unexpected order. i.e. before subscriptions always applied to the base layer while now they are also queued up as described [here](https://formidable.com/open-source/urql/docs/graphcache/normalized-caching/#deterministic-cache-updates)
